collector/dhcp.go

```diff
@@ -0,0 +1,412 @@
+	if desc, err := c.collect(ctx, ch); err != nil {
+		log.Error("failed collecting dhcp metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+// represents perflib metrics from the DHCP Server class.
+// While the name of a number of perflib metrics would indicate a rate is being returned (E.G. Packets Received/sec),
+// perflib instead returns a counter, hence the "Total" suffix in some of the variable names.
+type dhcpPerf struct {
+	PacketsReceivedTotal                             float64 `perflib:"Packets Received/sec"`
+	DuplicatesDroppedTotal                           float64 `perflib:"Duplicates Dropped/sec"`
+	PacketsExpiredTotal                              float64 `perflib:"Packets Expired/sec"`
+	MillisecondsPerPacketAvg                         float64 `perflib:"Milliseconds per packet (Avg)."`
+	ActiveQueueLength                                float64 `perflib:"Active Queue Length"`
+	ConflictCheckQueueLength                         float64 `perflib:"Conflict Check Queue Length"`
+	DiscoversTotal                                   float64 `perflib:"Discovers/sec"`
+	OffersTotal                                      float64 `perflib:"Offers/sec"`
+	RequestsTotal                                    float64 `perflib:"Requests/sec"`
+	InformsTotal                                     float64 `perflib:"Informs/sec"`
+	AcksTotal                                        float64 `perflib:"Acks/sec"`
+	NacksTotal                                       float64 `perflib:"Nacks/sec"`
+	DeclinesTotal                                    float64 `perflib:"Declines/sec"`
+	ReleasesTotal                                    float64 `perflib:"Releases/sec"`
+	OfferQueueLength                                 float64 `perflib:"Offer Queue Length"`
+	DeniedDueToMatch                                 float64 `perflib:"Denied due to match."`
+	DeniedDueToNonmatch                              float64 `perflib:"Denied due to nonmatch."`
+	FailoverBndupdSentTotal                          float64 `perflib:"Failover: BndUpd sent/sec."`
+	FailoverBndupdReceivedTotal                      float64 `perflib:"Failover: BndUpd received/sec."`
+	FailoverBndackSentTotal                          float64 `perflib:"Failover: BndAck sent/sec."`
+	FailoverBndackReceivedTotal                      float64 `perflib:"Failover: BndAck received/sec."`
+	FailoverBndupdPendingOutboundQueue               float64 `perflib:"Failover: BndUpd pending in outbound queue."`
+	FailoverTransitionsCommunicationinterruptedState float64 `perflib:"Failover: Transitions to COMMUNICATION-INTERRUPTED state."`
+	FailoverTransitionsPartnerdownState              float64 `perflib:"Failover: Transitions to PARTNER-DOWN state."`
+	FailoverTransitionsRecoverState                  float64 `perflib:"Failover: Transitions to RECOVER state."`
+	FailoverBndupdDropped                            float64 `perflib:"Failover: BndUpd Dropped."`
```

Comment:
These also sound counter-y, or how should a gauge value be interpreted?

Reply:
I think `FailoverBndupdDropped` is likely a counter, but the other metrics may be indicating the DHCP server failover state. E.G `FailoverTransitionsPartnerdownState 1` may indicate the DHCP server's partner is down, or that 1 partner in the failover cluster is down.

I'll see if I can find more documentation on this, as forcing a partner down in our prod environment may not win me any friends :sweat_smile:

Reply:
"Ooops" :smile:

Reply:
After monitoring, I've seen the `FailoverTransitionsCommunicationinterruptedState` metric increment from 51 to 55 on the primary, and 11 to 15 on the secondary. I think it's safe to assume this is a counter.
I would also assume the `FailoverTransitionsPartnerdownState` and `FailoverTransitionsRecoverState` are also counters. I'm less certain of `FailoverBndupdDropped` but am comfortable leaving it as a counter.